modules/sfp_grayhatwarfare.py

```diff
@@ -0,0 +1,169 @@
+#
+# Author:      <krishnasis@hotmail.com>
+#
+# Created:     24-01-2021
+# Copyright:   (c) Steve Micallef
+# Licence:     GPL
+# -------------------------------------------------------------------------------
+
+import json
+import time
+from spiderfoot import SpiderFootEvent, SpiderFootPlugin
+
+
+class sfp_grayhatwarfare(SpiderFootPlugin):
+
+    meta = {
+        'name': "Grayhat Warfare",
+        'summary': "Find bucket names matching the keyword extracted from a domain from Grayhat API.",
+        'flags': ["apikey"],
+        'useCases': ["Footprint", "Investigate", "Passive"],
+        'categories': ["Reputation Systems"],
+        'dataSource': {
+            'website': "https://buckets.grayhatwarfare.com/",
+            'model': "FREE_AUTH_LIMITED",
+            'references': [
+                "https://buckets.grayhatwarfare.com/docs/api/v1"
+            ],
+            'favIcon': "https://buckets.grayhatwarfare.com/assets/template/images/favicon.png",
+            'logo': "https://buckets.grayhatwarfare.com/assets/images/logo/logo-sm.png",
+            'description': "It is a searchable database of open buckets."
+            "Has up to million results of each bucket."
+            "Full text search with binary logic (can search for keywords and also stopwords)",
+        }
+    }
+
+    # Default options
+    opts = {
+        'api_key': '',
+        'per_page': 1000,
+        'max_pages': 2,
+        'pause': 1
+    }
+
+    # Option descriptions
+    optdescs = {
+        'api_key': 'Grayhat Warfare API key.',
+        'per_page': 'Maximum number of results per page (Max: 1000).',
+        'max_pages': 'Maximum number of pages to fetch.',
+        'pause': 'Number of seconds to wait between each API call.'
+    }
+
+    results = None
+    errorState = False
+
+    def setup(self, sfc, userOpts=dict()):
+        self.sf = sfc
+        self.results = self.tempStorage()
+
+        for opt in list(userOpts.keys()):
+            self.opts[opt] = userOpts[opt]
+
+    # What events is this module interested in for input
+    def watchedEvents(self):
+        return [
+            "DOMAIN_NAME",
+        ]
+
+    # What events this module produces
+    def producedEvents(self):
+        return [
+            'CLOUD_STORAGE_BUCKET',
+            'CLOUD_STORAGE_BUCKET_OPEN',
+            'RAW_RIR_DATA'
+        ]
+
+    # Query Grayhat Warfare
+    def query(self, keyword, start):
+        queryString = f"https://buckets.grayhatwarfare.com/api/v1/buckets/{start}/{self.opts['per_page']}?access_token={self.opts['api_key']}&keywords={keyword}"
+
+        headers = {
+            'Accept': 'application/json',
+        }
+
+        res = self.sf.fetchUrl(queryString,
+                               headers=headers,
+                               timeout=15,
+                               useragent=self.opts['_useragent'],
+                               verify=True)
```

Comment:
The usual code pattern is to `import urllib` and use `urllib.parse.urlencode` to prevent issues with encoding.

```suggestion
        params = urllib.parse.urlencode({
            'keywords': keyword.encode('raw_unicode_escape'),
            'access_token': self.opts['api_key']
        })

        headers = {
            'Accept': 'application/json',
        }

        res = self.sf.fetchUrl(
            f"https://buckets.grayhatwarfare.com/api/v1/buckets/{start}/{self.opts['per_page']}?{params}",
            headers=headers,
            timeout=15,
            useragent=self.opts['_useragent'],
            verify=True
        )
```